src/blocks/post-carousel/index.php

```diff
@@ -341,3 +341,35 @@ function coblocks_register_post_carousel_block() {
 	);
 }
 add_action( 'init', 'coblocks_register_post_carousel_block' );
+
+
+
+/**
+ * Handles outdated versions of the `coblocks/post-carousel` block by converting
+ * attribute `categories` from a numeric string to an array with key `id`.
+ *
+ * This is done to accommodate the changes introduced in https://github.com/WordPress/gutenberg/pull/20781 that sought to
+ * add support for multiple categories to the block. However, given that this
+ * block is dynamic, the usual provisions for block migration are insufficient,
+ * as they only act when a block is loaded in the editor.
+ *
+ * TODO: Remove when and if the bottom client-side deprecation for this block
+ * is removed.
+ *
+ * @param array $block A single parsed block object.
+ *
+ * @return array The migrated block object.
+ */
+function block_coblocks_post_carousel_migrate_categories( $block ) {
+	if (
+		'coblocks/post-carousel' === $block['blockName'] &&
+		! empty( $block['attrs']['categories'] ) &&
+		is_string( $block['attrs']['categories'] )
+	) {
+		$block['attrs']['categories'] = array(
+			array( 'id' => absint( $block['attrs']['categories'] ) ),
+		);
+	}
+	return $block;
+}
+add_filter( 'render_block_data', 'block_coblocks_post_carousel_migrate_categories' );
```

Comment:
I don't think we actually need to do this. Instead we should handle the existence of both keys and manage it within `coblocks_render_post_carousel_block` which I believe you already are.

Reply:
You are saying that I should not use the filter and instead place this logic elsewhere? Within the function `coblocks_render_post_carousel_block`? 

The logic to migrate from string type attribute to array type attribute does not presently exist meaning that any configured categories do **not** function correctly after migration.

Reply:
@jrtashjian  This filter will handle the migration on published pages without the need for the user to enter the editor and update markup. As a result, without the filter, previously configured and saved blocks that once properly filtered by category on the front-end no longer function as expected. This filter is for that specific case. 

Otherwise, you are correct that the deprecated migration function handles the migration type transform from string to an array of objects. This **does** require a user to enter the editor, however.

Reply:
With Filter Works
![image](https://user-images.githubusercontent.com/30462574/91741161-2995d480-eb69-11ea-83f0-ce1864a3c845.png)

Without Filter Broken
![withoutPHPFilterbroken](https://user-images.githubusercontent.com/30462574/91741850-27804580-eb6a-11ea-8ac8-ad7db58a4e1b.gif)

Reply:
A bit more about the need for the attribute migration that is not handled in https://github.com/godaddy-wordpress/coblocks/blob/67f84b3e1911fddcfcba23c184b62dc3bfd4c40b/src/blocks/post-carousel/index.php#L28-L32

https://github.com/godaddy-wordpress/coblocks/blob/67f84b3e1911fddcfcba23c184b62dc3bfd4c40b/src/blocks/post-carousel/block.json#L63-L68
```
<!-- wp:coblocks/post-carousel {"categories":"4"} /-->
```
I suspect that due to changes within the block.json attribute schema the classic attribute syntax saved into post content is not properly parsed into the new format. In this case above the `$attributes['categories']` is unset and treated as undefined. The filter accomplishes the goal of correctly parsing the attribute from post content thus allowing migration from categories attribute typed string to categories attribute typed as an array of objects.

Reply:
Marking this resolved since the original solution is sufficient.